compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -2514,23 +2496,6 @@ fn add_upstream_rust_crates<'a>(
         );
     }
 
-    // compiler-builtins are always placed last to ensure that they're
-    // linked correctly.
-    // We must always link the `compiler_builtins` crate statically. Even if it
-    // was already "included" in a dylib (e.g., `libstd` when `-C prefer-dynamic`
-    // is used)
-    if let Some(cnum) = compiler_builtins {
-        add_static_crate(
-            cmd,
-            sess,
-            archive_builder_builder,
-            codegen_results,
-            tmpdir,
-            cnum,
-            &Default::default(),
-        );
-    }
```

Comment:
This has to be placed after everything, not just after rust crates but also after native dependencies in case libgcc_s or libcompiler_rt aren't linked.

Reply:
Compiler builtins are placed after all crates and all native libs those crates bring.
It's not something this PR changes, it's the behavior both before and after this PR.

Reply:
I can't find the code that causes compiler_builtins to be placed after all native libs.

Reply:
What if libgcc_s.so is linked? It should come before compiler_builtins to ensure that the intrinsics resolve in libgcc_s.so rather than libcompiler_builtins.rlib I think as the former is much more optimized.

Reply:
Also, is compiler-rt optimized on par with libgcc (I would expect so)?
If compiler_builtins is built with enabling C components borrowed from compiler-rt (https://docs.rs/crate/compiler_builtins/0.1.83/features#c, IIRC) then compiler_builtins should also be good in this regard.

Reply:
> But right now it's linked last, so I guess the versions of intrinsics from compiler_builtins are typically used on GNU targets (unless crt-static is enabled or something like that).

Right, I confirmed this just now. I was under the impression that compiler_builtins would be after dynamic libraries, but because this is not the case I will resolve this conversation.